relay-server/src/actors/project_upstream.rs

```diff
@@ -191,191 +230,240 @@ impl UpstreamProjectSource {
-                    no_cache: channels_batch.values().any(|c| c.no_cache),
-                };
-
-                // count number of http requests for project states
-                metric!(counter(RelayCounters::ProjectStateRequest) += 1);
-
-                UpstreamRelay::from_registry()
-                    .send(SendQuery(query))
-                    .map_err(|_| ProjectError::ScheduleFailed)
-                    .map(move |response| (channels_batch, response))
-            })
-            .collect();
+    /// Executes an upstream request to fetch project configs.
+    ///
+    /// This assumes that currently no request is running. If the upstream request fails or new
+    /// channels are pushed in the meanwhile, this will reschedule automatically.
+    async fn fetch_states(
+        config: Arc<Config>,
+        channels: ChannelsBatch,
+        inner_tx: mpsc::UnboundedSender<Vec<UpstreamResponse>>,
+    ) {
+        let request_start = Instant::now();
+        let batch_size = config.query_batch_size();
+        let cache_batches = channels.cache_channels.into_iter().chunks(batch_size);
+        let nocache_batches = channels.nocache_channels.into_iter().chunks(batch_size);
+
+        let mut requests = vec![];
+        for channels_batch in cache_batches.into_iter().chain(nocache_batches.into_iter()) {
+            let mut channels_batch: ProjectStateChannels = channels_batch.collect();
+            for channel in channels_batch.values_mut() {
+                channel.attempts += 1;
+            }
+            relay_log::debug!("sending request of size {}", channels_batch.len());
+            metric!(
+                histogram(RelayHistograms::ProjectStateRequestBatchSize) =
+                    channels_batch.len() as u64
+            );
+
+            let query = GetProjectStates {
+                public_keys: channels_batch.keys().copied().collect(),
+                full_config: config.processing_enabled(),
+                no_cache: channels_batch.values().any(|c| c.no_cache),
+            };
+
+            // count number of http requests for project states
+            metric!(counter(RelayCounters::ProjectStateRequest) += 1);
+
+            let future_request = compat::send(UpstreamRelay::from_registry(), SendQuery(query))
+                .map(|response| match response {
+                    Ok(response) => UpstreamResponse {
+                        channels_batch,
+                        response: response.map_err(ProjectError::UpstreamFailed),
+                    },
+                    // Also propagate the channels in error case, since we will want to try to
+                    // fetch the projects again.
+                    _ => UpstreamResponse {
+                        channels_batch,
+                        response: Err::<GetProjectStatesResponse, ProjectError>(
+                            ProjectError::ScheduleFailed,
+                        ),
+                    },
```

Comment:
This changes the behavior of the service. Unfortunately, the prior code was not entirely clear on this. Particularly `ProjectError::ScheduleFailed` was actually dead code (check it by removing `.map_err(|_| ProjectError::ScheduleFailed)` from master). 

There are two kinds of error that can happen:
 1. We fail to send a message to the upstream. This error is propagated through `join_all.and_then()`, so the handler doesn't run. In this case, we back off immediately and drop the affected channels. This, in turn, will automatically fail the project config request in the `ProjectCache`. We used to map this to `ProjectError::ScheduleFailed`, but this error was immediately dropped. So this error can be removed entirely.
 2. The upstream responds with an error. That went into a match statement, which resulted in a log and putting the channels back. 

Put together, I think we can remove `ProjectError` and simply return an `UpstreamError` here.

Reply:
> so the handler doesn't run. In this case, we back off immediately and drop the affected channels. 
So this also means we do not schedule fetches anymore, in case of the error `#1`:
* we never trigger `schedule_fetch`
* and because of https://github.com/getsentry/relay/blob/2a0976d7d9ecfd4daa6cb2390c224d09f9a79bbd/relay-server/src/actors/project_upstream.rs#L348 we will never trigger it on new message either

Do I see it correctly or am I missing something obvious?